### DISCUSSION TITLE
Generating constraints for usage in errors/conflicts

### DISCUSSION THREAD

**Discussion by daddz** [stephenafamo/bob#242] [Ideas]
I saw this nice PR: https://github.com/stephenafamo/bob/pull/237 and it sparked an idea but I don't know if it's feasible.

Would it be possible to also generate variables/fields for the (named?) constraints on a table?

Bogus example:
```sql
ALTER TABLE pilots ADD CONSTRAINT uniq_email UNIQUE (email);
ALTER TABLE pilots ADD CONSTRAINT multi_field_example UNIQUE (field1, field2);
```

Would generate some fields like this:
```go
models.Pilots.Constraints.UniqEmail
models.Pilots.Constraints.MultiFieldExample
```

Which could be checked against in error cases:
```go
if _, err := models.Pilots.Insert(...); err != nil {
  if models.Pilots.Constraints.UniqEmail.IsErr(err) {
    fmt.Println("email must be unique")
  }
  // alternative api idea
  if constraints.IsErr(models.Pilots.Constraints.UniqEmail, err) {
  // ...
  }
}
```

Maybe it could also be used in e.g. conflicts for `Upsert` for constraints that span multiple columns:
```go
models.Pilots.Upsert(ctx, db, true, []string{models.Pilots.Constraints.MultiFieldExample}, ...)
```

I think something like this could greatly enhance error and conflict handling.

I didn't look into the PR in detail yet so the generation part might be solved/available already

**Comment by stephenafamo** [comment]
This would be nice.

The primary bottleneck will be reading the constraint information from each DB. Once that is done, generating code would be relatively straightforward.

**Comment by daddz** [comment]
@mbezhanov I saw you have some plans that (I think) are going in this direction (https://github.com/stephenafamo/bob/pull/237#issuecomment-2173868731)

What are your thoughts on this?

**Comment by mbezhanov** [comment]
@daddz I'll look into this in the following days and make a proposal.

**Comment by mbezhanov** [comment]
@daddz @stephenafamo so I put some thought into this, and here's what I came up with.

As you know, the `errors` package in Go allows an `error` type to provide an `Is` method in order to be treated as equivalent to another `error` type. During code generation, we could create custom error types from the unique constraints in each table and then use these custom error types in conditional statements as follows:

```go
if _, err := models.Pilots.Insert(...); err != nil {
  if errors.Is(models.PilotErrors.ErrUniqueName, err) {
    // do something to handle the error
  }
}
```

To facilitate this, we could have the following type definition in `bob_main.go`:

```go
type errorUniqueConstraint struct {
  name string // the name of the unique constraint
}

func (e *errorUniqueConstraint) Error() string {
  return e.name
}

func (e *errorUniqueConstraint) Is(target error) bool {
  err, ok := target.(*mysql.MySQLError)
  if !ok {
    return false
  }
  return err.Number == 1062 && strings.Contains(err.Message, e.name)
}
```

Then, in the generated `pilots.go` file we could have something like this:

```go
type pilotErrors struct {
  ErrUniqueName error
}

var PilotErrors = pilotErrors{
  ErrUniqueName: &errorUniqueConstraint{name: "pilots_name_IDX"},
}
```

Obviously some of the hard-coded things (e.g. the constraint name, the MySQL error code, etc.) can be defined as constants and variables.

What do you think?

**Reply by stephenafamo** [reply]
This would probably provide the best developer experience.

However.... this requires that we know the exact driver being used:
* Postgres: `lib/pq` or `jackc/pgx`? If it is `pgx`, which version?
* For SQLite: `mattn/go-sqlite3` or `modernc.org/sqlite`?

**Reply by daddz** [reply]
@mbezhanov that is a really neat idea!

@stephenafamo do you think it would make sense to make the driver configurable in the generator? Might be a bit of an overhead to support different drivers but realistically the number of widely used drivers shouldn't grow too much if at all.

Not sure if it would be possible to use the raw errors from `database/sql` and parse them directly (maybe something like https://github.com/jackc/pgerrcode would help).

**Reply by stephenafamo** [reply]
I'm not sure there's a way to use the raw errors from `database/sql` since the errors returned would be given by the driver.

I agree that supporting different drivers would be the best way, but then we have to properly consider how it would work so that it has great presets for the common drivers, but easy to override for custom drivers.

**Reply by mbezhanov** [reply]
An alternative way to get around the type assertion and the need to know which particular driver to use could be to perform a simple substring search and/or regular expression matching (i.e., `strings.Contains()` and/or `regexp.MatchString()`).

I had a look at `pq` and `pgx`, and both of them seem to include the raw error message string returned from the database in their error reports:

```go
// pgx v3
// ERROR: duplicate key value violates unique constraint "pilots_pkey" (SQLSTATE 23505)
func (pe PgError) Error() string {
  return pe.Severity + ": " + pe.Message + " (SQLSTATE " + pe.Code + ")"
}

// pgx v4
// ERROR: duplicate key value violates unique constraint "pilots_pkey" (SQLSTATE 23505)
func (pe *PgError) Error() string {
  return pe.Severity + ": " + pe.Message + " (SQLSTATE " + pe.Code + ")"
}

// pgx v5
// duplicate key value violates unique constraint "pilots_pkey" (SQLSTATE 23505)
func (pe *PgError) Error() string {
  return pe.Severity + ": " + pe.Message + " (SQLSTATE " + pe.Code + ")"
}

// pq
// pq: duplicate key value violates unique constraint "pilots_pkey"
func (err *Error) Error() string {
  return "pq: " + err.Message
}
```

It's obviously not ideal to rely on an error message string to do the matching, yet digging into the Postgres source code I found out that the string `duplicate key value violates unique constraint "%s"` hasn't changed since version `8.3.0`. That version was released back in 2008, which gives some level of confidence that this error message format is likely to remain consistent in future versions of Postgres as well.

Complete evolution of the message listed below:

| Commit date  | Message format |
| ------------- | ------------- |
| 04/06/2007  | `duplicate key value violates unique constraint \"%s\"` |
| 25/09/2003  | `duplicate key violates unique constraint \"%s\"` |
| 21/07/2003  | `duplicate key violates UNIQUE constraint \"%s\"` |
| 22/11/1999  | `Cannot insert a duplicate key into a unique index %s` |
| 10/01/1997 | `Cannot insert a duplicate key into a unique index` |

---

With MySQL, things are a little trickier because error messages can be localized. In other words, the duplicate key error message could take any of the following forms:

```txt
ER_DUP_ENTRY 23000 S1009
  cze "Zdvojený klíč '%-.192s' (číslo klíče %d)"
  dan "Ens værdier '%-.192s' for indeks %d"
  nla "Dubbele ingang '%-.192s' voor zoeksleutel %d"
  eng "Duplicate entry '%-.192s' for key %d"
  est "Kattuv väärtus '%-.192s' võtmele %d"
  fre "Duplicata du champ '%-.192s' pour la clef %d"
  ger "Doppelter Eintrag '%-.192s' für Schlüssel %d"
  greek "Διπλή εγγραφή '%-.192s' για το κλειδί %d"
  hun "Duplikalt bejegyzes '%-.192s' a %d kulcs szerint."
  ita "Valore duplicato '%-.192s' per la chiave %d"
  jpn "'%-.192s' は索引 %d で重複しています。"
  kor "중복된 입력 값 '%-.192s': key %d"
  nor "Like verdier '%-.192s' for nøkkel %d"
  norwegian-ny "Like verdiar '%-.192s' for nykkel %d"
  pol "Powtórzone wystąpienie '%-.192s' dla klucza %d"
  por "Entrada '%-.192s' duplicada para a chave %d"
  rum "Cimpul '%-.192s' e duplicat pentru cheia %d"
  rus "Дублирующаяся запись '%-.192s' по ключу %d"
  serbian "Dupliran unos '%-.192s' za ključ '%d'"
  slo "Opakovaný kľúč '%-.192s' (číslo kľúča %d)"
  spa "Entrada duplicada '%-.192s' para la clave %d"
  swe "Dublett '%-.192s' för nyckel %d"
  ukr "Дублюючий запис '%-.192s' для ключа %d"
```

The good thing, though, is that the `go-sql-driver` reports both the message and the error code, which would allow us to do the matching:

```go
func (me *MySQLError) Error() string {
  if me.SQLState != [5]byte{} {
    return fmt.Sprintf("Error %d (%s): %s", me.Number, me.SQLState, me.Message)
  }
  
  return fmt.Sprintf("Error %d: %s", me.Number, me.Message)
}
```
---

I didn't check the SQLite drivers, but I suppose something similar could be possible there as well.

---

Do you think a similar solution is worth implementing, or would it be too fragile and unreliable?

**Reply by stephenafamo** [reply]
I'm conflicted...

On one hand, this makes it driver-independent.
It is also probably easier to customize for a user because they only need to provide a different regexp to match.

On the other hand, string matching for errors feels awkward.

**Reply by daddz** [reply]
Having something driver-independent would be great but I don't think it would be a nice/maintainable  due to the differences in error reporting.

If the stdlib `database/sql` would provide better abstractions for error reporting from drivers then it would be a different story maybe.

Based on the assumption that the current driver landscape is pretty much stable it might be worth to think about e.g. supporting only the 1-2 most-widely used ones for each database.

And in order to make things simple, one might start with driver-specific code-generation only for the error parts. That way the whole driver-specific part should be fairly small and it's simpler to try out different concepts/ideas on how to do that in general.

Maybe the string matching could be used as a fall-back if a driver is not supported?

If that works out nicely it could later be extended to other parts that might benefit from driver-specific code generation.

**Reply by stephenafamo** [reply]
I think the best way would be to do it this way:

1. Have driver specific code:
    * Similar to how you can select your `uuid` package with the postgres generator, you can select your connection driver
    * Allow configuration for adding custom drivers.
    * To prevent surprises, perhaps we require the driver to be specified and either have an error if none is specified, or skip the driver specific code
2. Allow users to setup custom drivers. Either through configuration or templates.

**Reply by mbezhanov** [reply]
Sounds good. I'll work on preparing a PR, aiming to support the drivers you mentioned earlier.

**Reply by mbezhanov** [reply]
I prepared a draft PR over here to propose an implementation: https://github.com/stephenafamo/bob/pull/248
